Generalize dynamic flag

lucaswerkmeister opened the pull request:
This moves the `dynamic` flag from `TypeDeclaration` to `Declaration` and removes the `jsNew` flag.

@chochos I was working on this in an older branch (before NPM support), and was surprised to discover that the backend still correctly emitted `new` for dynamic classes after I removed `jsNew` – the backend already knew how to treat dynamic classes correctly, which I wasn’t aware of when I wrote #6050. After I rebased on `master`, this broke, because you removed [this](https://github.com/ceylon/ceylon/blob/5c89be787ff7dc07ec527c44e2ce8e71dd131f33/compiler-js/src/main/java/com/redhat/ceylon/compiler/js/BmeGenerator.java#L267-L269) bit in 9bb42bb. I’ve added it back (removing [this](https://github.com/ceylon/ceylon/blob/a1a79d5a93046132704f87af2ada175306b9ca4b/compiler-js/src/main/java/com/redhat/ceylon/compiler/js/InvocationGenerator.java#L711-L725) from #6050), because the old version looks better to me. Do you agree?

Also, @gavinking might want to glance over the model part of the changes.

lucaswerkmeister committed:
Update TypedDeclaration flag name and documentation

lucaswerkmeister committed:
Move dynamic flag from TypeDeclaration to Declaration

With TypeScript support, toplevel values and functions also need to be
considered dynamic (no getters/setters, no reified types, no escaped
names, etc.), so the flag needs to be more general. (This is also not
the same as the dynamicallyTyped flag – a trap I had been walking into
for a while before realizing my mistake.)

lucaswerkmeister committed:
Remove jsNew flag (superseded by dynamic flag)

Now that the dynamic flag is more general, it can include the meaning of
the jsNew flag, which is no longer needed.

Reverts #6050 and part of 9bb42bb.

lucaswerkmeister commented:
This also opens another question. Before this change, there’s no difference between members of a dynamic interface and members of a regular interface – they’re treated differently because their *container* is dynamic. Now that we can also make the members themselves dynamic, we could change that: emit a `dyn` flag for those members, and check `member.isDynamic()` instead of `container.isDynamic()`. On the other hand, that’s another binary compatibility change (though not a breaking one – when loading a dynamic interface from `9.0` instead of `9.1`, add the dynamic flag to its members).

chochos commented:
But the members of a dynamic interface are themselves already dynamic – they are typed, yes, but the only check at runtime is that they exist, because the type obviously can't be checked. And regarding members that are themselves types, I don't see how a dynamic interface could have an inner class that is not itself dynamic.

Or, are you talking about checking `member.isDynamic()`, but the implementation of that inside `Function` and `Value` would be to check if its container is dynamic?

lucaswerkmeister commented:
> Or, are you talking about checking `member.isDynamic()`, but the implementation of that inside `Function` and `Value` would be to check if its container is dynamic?

Yes, I mean `member.isDynamic()` (which simply didn’t exist before), but I don’t think `Function` and `Value` should do anything special – we could just make them dynamic in the `-model.js` file and teach the model loader to load that.

lucaswerkmeister committed:
Store dynamic flag for methods and attributes

lucaswerkmeister committed:
Add dynamic flag for members of dynamic interfaces

lucaswerkmeister committed:
Add dynamic flag for members of NPM classes

lucaswerkmeister committed:
Read dynamic flag of members instead of their containers

Now that we have the dynamic flag on typed declarations as well (not
only on type declarations), we don't have to look at the container to
determine if a function or value should be treated like a dynamic one.

lucaswerkmeister committed:
Make members of old modules' dynamic interfaces dynamic

For new modules, we store the dynamic flag of dynamic interface members
(see previous commits), but for old modules, we have to add it manually.

lucaswerkmeister commented:
@chochos I’ve added some commits to implement that suggestion (backwards compatibility tested with `ceylon.interop.browser/1.2.2`).

lucaswerkmeister commented:
@chochos should I merge this or do you want to review it first? (I’ve got some more commits ready to support dynamic values… I would open a new pull request for that after merging this.)

lucaswerkmeister commented:
Alright, thanks!

lucaswerkmeister commented on compiler-js/src/main/java/com/redhat/ceylon/compiler/js/TypeGenerator.java:
```diff
@@ -295,8 +295,8 @@ static void interfaceDefinition(final Tree.InterfaceDefinition that, final Gener
-        //If it's inside a dynamic interface, don't generate anything
-        if (d.isClassOrInterfaceMember() && ((ClassOrInterface)d.getContainer()).isDynamic())return;
+        //If it's inside a dynamic interface (or for some other reason dynamic), don't generate anything
+        if (d.isDynamic())return;
```
This is complete rubbish, sorry. Fixed in 1d1a4c846318914cf897e62943531cbd2e21225a.
